// react-native-css/dist/commonjs/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ActivityIndicator: true,
  Button: true,
  FlatList: true,
  Image: true,
  ImageBackground: true,
  KeyboardAvoidingView: true,
  Pressable: true,
  ScrollView: true,
  Switch: true,
  Text: true,
  TextInput: true,
  TouchableHighlight: true,
  TouchableOpacity: true,
  TouchableWithoutFeedback: true,
  View: true,
  VirtualizedList: true
};
Object.defineProperty(exports, "ActivityIndicator", {
  enumerable: true,
  get: function () {
    return _ActivityIndicator.ActivityIndicator;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function () {
    return _Button.Button;
  }
});
Object.defineProperty(exports, "FlatList", {
  enumerable: true,
  get: function () {
    return _FlatList.FlatList;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function () {
    return _Image.Image;
  }
});
Object.defineProperty(exports, "ImageBackground", {
  enumerable: true,
  get: function () {
    return _ImageBackground.ImageBackground;
  }
});
Object.defineProperty(exports, "KeyboardAvoidingView", {
  enumerable: true,
  get: function () {
    return _KeyboardAvoidingView.KeyboardAvoidingView;
  }
});
Object.defineProperty(exports, "Pressable", {
  enumerable: true,
  get: function () {
    return _Pressable.Pressable;
  }
});
Object.defineProperty(exports, "ScrollView", {
  enumerable: true,
  get: function () {
    return _ScrollView.ScrollView;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function () {
    return _Switch.Switch;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _Text.Text;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function () {
    return _TextInput.TextInput;
  }
});
Object.defineProperty(exports, "TouchableHighlight", {
  enumerable: true,
  get: function () {
    return _TouchableHighlight.TouchableHighlight;
  }
});
Object.defineProperty(exports, "TouchableOpacity", {
  enumerable: true,
  get: function () {
    return _TouchableOpacity.TouchableOpacity;
  }
});
Object.defineProperty(exports, "TouchableWithoutFeedback", {
  enumerable: true,
  get: function () {
    return _TouchableWithoutFeedback.TouchableWithoutFeedback;
  }
});
Object.defineProperty(exports, "View", {
  enumerable: true,
  get: function () {
    return _View.View;
  }
});
Object.defineProperty(exports, "VirtualizedList", {
  enumerable: true,
  get: function () {
    return _VirtualizedList.VirtualizedList;
  }
});
var _reactNative = require("react-native");
Object.keys(_reactNative).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _reactNative[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _reactNative[key];
    }
  });
});
var _ActivityIndicator = require("./ActivityIndicator.js");
var _Button = require("./Button.js");
var _FlatList = require("./FlatList.js");
var _Image = require("./Image.js");
var _ImageBackground = require("./ImageBackground.js");
var _KeyboardAvoidingView = require("./KeyboardAvoidingView.js");
var _Pressable = require("./Pressable.js");
var _ScrollView = require("./ScrollView.js");
var _Switch = require("./Switch.js");
var _Text = require("./Text.js");
var _TextInput = require("./TextInput.js");
var _TouchableHighlight = require("./TouchableHighlight.js");
var _TouchableOpacity = require("./TouchableOpacity.js");
var _TouchableWithoutFeedback = require("./TouchableWithoutFeedback.js");
var _View = require("./View.js");
var _VirtualizedList = require("./VirtualizedList.js");
