// react-native-css/src/components/index.cts
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-require-imports */

module.exports = {
  get ActivityIndicator() {
    return require("./ActivityIndicator").ActivityIndicator;
  },
  get Button() {
    return require("./Button").Button;
  },
  get FlatList() {
    return require("./FlatList").FlatList;
  },
  get Image() {
    return require("./Image").Image;
  },
  get ImageBackground() {
    return require("./ImageBackground").ImageBackground;
  },
  get KeyboardAvoidingView() {
    return require("./KeyboardAvoidingView").KeyboardAvoidingView;
  },
  get Pressable() {
    return require("./Pressable").Pressable;
  },
  get ScrollView() {
    return require("./ScrollView").ScrollView;
  },
  get Switch() {
    return require("./Switch").Switch;
  },
  get Text() {
    return require("./Text").Text;
  },
  get TextInput() {
    return require("./TextInput").TextInput;
  },
  get TouchableHighlight() {
    return require("./TouchableHighlight").TouchableHighlight;
  },
  get TouchableOpacity() {
    return require("./TouchableOpacity").TouchableOpacity;
  },
  get View() {
    return require("./View").View;
  },
  get VirtualizedList() {
    return require("./VirtualizedList").VirtualizedList;
  },

  /**
   * We need to return
   */
  get DrawerLayoutAndroid() {
    return require("react-native").DrawerLayoutAndroid;
  },
  get InputAccessoryView() {
    return require("react-native").InputAccessoryView;
  },
  get experimental_LayoutConformance() {
    return require("react-native").experimental_LayoutConformance;
  },
  get Modal() {
    return require("react-native").Modal;
  },
  get ProgressBarAndroid() {
    return require("react-native").ProgressBarAndroid;
  },
  get RefreshControl() {
    return require("react-native").RefreshControl;
  },
  get SafeAreaView() {
    return require("react-native").SafeAreaView;
  },
  get SectionList() {
    return require("react-native").SectionList;
  },
  get StatusBar() {
    return require("react-native").StatusBar;
  },
  get unstable_TextAncestorContext() {
    return require("react-native").unstable_TextAncestorContext;
  },
  get Touchable() {
    return require("react-native").Touchable;
  },
  get TouchableNativeFeedback() {
    return require("react-native").TouchableNativeFeedback;
  },
  get TouchableWithoutFeedback() {
    return require("react-native").TouchableWithoutFeedback;
  },
  get VirtualizedSectionList() {
    return require("react-native").VirtualizedSectionList;
  },
  get unstable_VirtualView() {
    return require("react-native").unstable_VirtualView;
  },
  get AccessibilityInfo() {
    return require("react-native").AccessibilityInfo;
  },
  get ActionSheetIOS() {
    return require("react-native").ActionSheetIOS;
  },
  get Alert() {
    return require("react-native").Alert;
  },
  get Animated() {
    return require("react-native").Animated;
  },
  get Appearance() {
    return require("react-native").Appearance;
  },
  get AppRegistry() {
    return require("react-native").AppRegistry;
  },
  get AppState() {
    return require("react-native").AppState;
  },
  get BackHandler() {
    return require("react-native").BackHandler;
  },
  get Clipboard() {
    return require("react-native").Clipboard;
  },
  get codegenNativeCommands() {
    return require("react-native").codegenNativeCommands;
  },
  get codegenNativeComponent() {
    return require("react-native").codegenNativeComponent;
  },
  get DeviceEventEmitter() {
    return require("react-native").DeviceEventEmitter;
  },
  get DeviceInfo() {
    return require("react-native").DeviceInfo;
  },
  get DevMenu() {
    return require("react-native").DevMenu;
  },
  get DevSettings() {
    return require("react-native").DevSettings;
  },
  get Dimensions() {
    return require("react-native").Dimensions;
  },
  get DynamicColorIOS() {
    return require("react-native").DynamicColorIOS;
  },
  get Easing() {
    return require("react-native").Easing;
  },
  get findNodeHandle() {
    return require("react-native").findNodeHandle;
  },
  get I18nManager() {
    return require("react-native").I18nManager;
  },
  get InteractionManager() {
    return require("react-native").InteractionManager;
  },
  get Keyboard() {
    return require("react-native").Keyboard;
  },
  get LayoutAnimation() {
    return require("react-native").LayoutAnimation;
  },
  get Linking() {
    return require("react-native").Linking;
  },
  get LogBox() {
    return require("react-native").LogBox;
  },
  get NativeAppEventEmitter() {
    return require("react-native").NativeAppEventEmitter;
  },
  get NativeDialogManagerAndroid() {
    return require("react-native").NativeDialogManagerAndroid;
  },
  get NativeEventEmitter() {
    return require("react-native").NativeEventEmitter;
  },
  get NativeModules() {
    return require("react-native").NativeModules;
  },
  get Networking() {
    return require("react-native").Networking;
  },
  get PanResponder() {
    return require("react-native").PanResponder;
  },
  get PermissionsAndroid() {
    return require("react-native").PermissionsAndroid;
  },
  get PixelRatio() {
    return require("react-native").PixelRatio;
  },
  get Platform() {
    return require("react-native").Platform;
  },
  get PlatformColor() {
    return require("react-native").PlatformColor;
  },
  get PushNotificationIOS() {
    return require("react-native").PushNotificationIOS;
  },
  get processColor() {
    return require("react-native").processColor;
  },
  get registerCallableModule() {
    return require("react-native").registerCallableModule;
  },
  get requireNativeComponent() {
    return require("react-native").requireNativeComponent;
  },
  get RootTagContext() {
    return require("react-native").RootTagContext;
  },
  get Settings() {
    return require("react-native").Settings;
  },
  get Share() {
    return require("react-native").Share;
  },
  get StyleSheet() {
    return require("react-native").StyleSheet;
  },
  get Systrace() {
    return require("react-native").Systrace;
  },
  get ToastAndroid() {
    return require("react-native").ToastAndroid;
  },
  get TurboModuleRegistry() {
    return require("react-native").TurboModuleRegistry;
  },
  get UIManager() {
    return require("react-native").UIManager;
  },
  get unstable_batchedUpdates() {
    return require("react-native").unstable_batchedUpdates;
  },
  get useAnimatedValue() {
    return require("react-native").useAnimatedValue;
  },
  get useColorScheme() {
    return require("react-native").useColorScheme;
  },
  get useWindowDimensions() {
    return require("react-native").useWindowDimensions;
  },
  get UTFSequence() {
    return require("react-native").UTFSequence;
  },
  get Vibration() {
    return require("react-native").Vibration;
  },
  get VirtualViewMode() {
    return require("react-native").VirtualViewMode;
  },
};
